language/tests/regent/compile_fail/optimize_index_launch_noninjective.rg

```diff
@@ -0,0 +1,34 @@
+-- Copyright 2021 Stanford University
+--
+-- Licensed under the Apache License, Version 2.0 (the "License");
+-- you may not use this file except in compliance with the License.
+-- You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+-- fails-with:
+-- Errors reported during runtime.
+-- optimize_index_launch_noninjective.rg:30: loop ineligible for index launch due to non-injective projection functor
```

Comment:
Do we want necessarily want `N` in the error message? This might be slightly tricky to implement since we don't know it's value at compile time. I'll take a look.

Reply:
Don't you know which argument you're generating the dynamic check for? To be clear, `N` is the position of the parameter, not the index of the color in the partition that caused the problem.

Reply:
I do, but I don't know which check failed if there are multiple non-trivial projection functors. The way the code is structured right now, at the point of issuing the `assert_error` call, I just know that at least one check failed (but not which one).

Reply:
Ok. It's not a huge deal if you want to leave this, but it would be nice to have.

Either way is fine, but if you want to implement it, I think the way to handle this would be to add a new variable `failing_argument` that you can initialize to `-1` (or something) and then set when a parameter check fails. But then (and this is the slightly ugly part) you'd need to basically inline the `assert_error` code into your generated code here because that function isn't quite general enough to handle the arbitrary format string required to interpolate the argument number. (Maybe there's a way to make `assert_error` a Terra macro, but at this point we're getting far enough out into the weeds that we should probably just focus on what makes it work.)

Reply:
I refactored the check to throw the error at a point where we do know which argument we're looking at. Now, in the demand case, we throw a runtime error as soon as we find a bad projection functor.